### PR TITLE
Parse cql files with code comments

### PR DESCRIPTION
**Pull request by ssserj**
CqlFileParser reimplemented (#56). We should ignore code comments, normalize queries and preserve newline in valueExpressions.

### CODE REVIEW

**Comment by awinder** [comment]
FWIW: I just tested this on a bunch of existing cql migration files (~50?) with comment lines and it worked swimmingly, the only problem I encountered was parsing multiple empty lines at the end of the file created an error:

```
Exception in thread "main" java.lang.IllegalStateException: File had a non-terminated cql line
       	at com.google.common.base.Preconditions.checkState(Preconditions.java:173)
       	at uk.sky.cqlmigrate.CqlFileParser$LineProcessor.check(CqlFileParser.java:153)
       	at uk.sky.cqlmigrate.CqlFileParser$LineProcessor.access$000(CqlFileParser.java:50)
       	at uk.sky.cqlmigrate.CqlFileParser.getCqlStatementsFrom(CqlFileParser.java:45)
       	at uk.sky.cqlmigrate.SchemaLoader$Loader.apply(SchemaLoader.java:44)
       	at uk.sky.cqlmigrate.CqlPaths.lambda$applyInSortedOrder$3(CqlPaths.java:56)
       	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
       	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
       	at java.util.Iterator.forEachRemaining(Iterator.java:116)
       	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
       	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
       	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
       	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
       	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
       	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
       	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
       	at uk.sky.cqlmigrate.CqlPaths.applyInSortedOrder(CqlPaths.java:56)
       	at uk.sky.cqlmigrate.SchemaLoader.load(SchemaLoader.java:28)
       	at uk.sky.cqlmigrate.CqlMigratorImpl.migrate(CqlMigratorImpl.java:95)
       	at uk.sky.cqlmigrate.CqlMigratorImpl.migrate(CqlMigratorImpl.java:67)
       	at uk.sky.cqlmigrate.CqlMigratorImpl.main(CqlMigratorImpl.java:58)
```

(this was an easy enough fix, just figured I'd mention as the only thing that we went wrong).  It'd be great if this PR merged, I think having support for commenting in CQL migrations would be a nice feature 😄

File: src/main/java/uk/sky/cqlmigrate/CqlFileParser.java [hunk 1]
```diff
@@ -22,37 +22,139 @@ private CqlFileParser() {}
 
     private static final char CQL_STATEMENT_STRING_DELIMITER = '\'';
     private static final String CQL_STATEMENT_TERMINATOR = ";";
+    private static final String CQL_COMMENT_DOUBLE_HYPEN = "--"; //Double hypen
+    private static final String CQL_MULTI_LINE_COMMENT_OPEN = "/*"; //Forward slash asterisk
+    private static final String CQL_MULTI_LINE_COMMENT_CLOSE = "*/"; //Asterisk forward slash
```
Comment (adamdougal) [inline src/main/java/uk/sky/cqlmigrate/CqlFileParser.java]: I'm not sure the comments for the 3 constants above add any value.

File: src/main/java/uk/sky/cqlmigrate/CqlFileParser.java [hunk 2]
```diff
@@ -22,37 +22,139 @@ private CqlFileParser() {}
...
+    private static final Pattern CQL_MULTI_LINE_COMMENT_PATTERN = Pattern.compile("/\\*.*?\\*/", Pattern.DOTALL);
+    private static final Pattern EOL = Pattern.compile(".*\\R|.+\\z");
+    private static final String EMPTY_STR = "";
```
Comment (adamdougal) [inline src/main/java/uk/sky/cqlmigrate/CqlFileParser.java]: All of these constants could move into the LineProcessor class.

File: src/main/java/uk/sky/cqlmigrate/CqlFileParser.java [hunk 3]
```diff
@@ -22,37 +22,139 @@ private CqlFileParser() {}
...
 
     static List<String> getCqlStatementsFrom(Path cqlPath) {
-        String cqlFileAsString;
+        final LineProcessor processor = new LineProcessor();
```
Comment (adamdougal) [inline src/main/java/uk/sky/cqlmigrate/CqlFileParser.java]: Unfortunately we don't have any code style guidelines, but normally we wouldn't use the `final` keyword for variables.

File: src/main/java/uk/sky/cqlmigrate/CqlFileParser.java [hunk 4]
```diff
@@ -22,37 +22,139 @@ private CqlFileParser() {}
...
-                candidateStatement = "";
-            } else {
-                candidateStatement += CQL_STATEMENT_TERMINATOR;
+    static class LineProcessor {
+        private static final byte INIT = 0;
+        private static final byte FIND_EOS = 1;
+        private static final byte IS_MULTI_LINE_COMMENT = 2;
+        private static final byte IS_OPEN_STMT = 3;
+        private static final byte IS_OPEN_VALUE_EXP = 4;
+        private static final byte IS_CLOSE_STMT = 5;
```
Comment (adamdougal) [inline src/main/java/uk/sky/cqlmigrate/CqlFileParser.java]: I'm wondering if a `LineProcessorState` enum might be better to use rather then constants.

File: src/main/java/uk/sky/cqlmigrate/CqlFileParser.java [hunk 5]
```diff
@@ -22,37 +22,139 @@ private CqlFileParser() {}
...
+        String[] statements;
+        int index;
+        byte curState = INIT;
+        StringBuilder curStmt;
+
+        void process(String original) throws IOException {
+            switch (curState) {
+                case INIT:
+                    if (statements == null) {
+                        statements = new String[1];
```
Comment (adamdougal) [inline src/main/java/uk/sky/cqlmigrate/CqlFileParser.java]: Is there any benefit to using an array rather then a list?

File: src/main/java/uk/sky/cqlmigrate/CqlFileParser.java [hunk 6]
```diff
@@ -22,37 +22,139 @@ private CqlFileParser() {}
...
+                case IS_MULTI_LINE_COMMENT:
+                    if (CharMatcher.WHITESPACE.trimTrailingFrom(original).endsWith(CQL_MULTI_LINE_COMMENT_CLOSE))
+                        curState = FIND_EOS;
+
+                    break;
+
+                case IS_CLOSE_STMT:
+                    LOGGER.trace("CQL parsed: {}", original);
+                    curState = INIT;
+                    break;
```
Comment (adamdougal) [inline src/main/java/uk/sky/cqlmigrate/CqlFileParser.java]: This method is quite big, can it be broken up into smaller private methods?

File: src/main/java/uk/sky/cqlmigrate/CqlFileParser.java [hunk 7]
```diff
@@ -22,37 +22,139 @@ private CqlFileParser() {}
...
-
-        for (String statementFragment : cqlStatements.split(CQL_STATEMENT_TERMINATOR)) {
-            candidateStatement += statementFragment;
-            // A semicolon preceded by an odd number of single quotes must be within a string, and therefore is not a statement terminator
-            if (CharMatcher.is(CQL_STATEMENT_STRING_DELIMITER).countIn(candidateStatement) % 2 == 0) {
-                statementList.add(candidateStatement);
-                candidateStatement = "";
-            } else {
-                candidateStatement += CQL_STATEMENT_TERMINATOR;
+    static class LineProcessor {
```
Comment (adamdougal) [inline src/main/java/uk/sky/cqlmigrate/CqlFileParser.java]: Can this be private?

File: src/test/java/uk/sky/cqlmigrate/CqlFileParserTest.java [hunk 8]
```diff
@@ -30,4 +32,66 @@ public void shouldCopeWithSemicolonsInStrings() throws Exception {
...
     }
+
+    @Test
+    public void shouldIgnoreCommentsAndNormolizeAndPreserveNewlineInValuesExp() throws Exception {
+        //given
+        Path cqlPath = getResourcePath("cql_rolegraphs_one/2015-08-16-12:05-statement-with-comments.cql");
+
+        //when
+        List<String> cqlStatements = CqlFileParser.getCqlStatementsFrom(cqlPath);
+        //cqlStatements.forEach(s -> System.out.println("CQL:" + s + "\n"));
```
Comment (adamdougal) [inline src/test/java/uk/sky/cqlmigrate/CqlFileParserTest.java]: Is this still needed?

File: src/test/java/uk/sky/cqlmigrate/CqlFileParserTest.java [hunk 9]
```diff
@@ -30,4 +32,66 @@ public void shouldCopeWithSemicolonsInStrings() throws Exception {
...
+                " VALUES ('EARTH', '', '   the end ')";
+
+        assertEquals(cqlStatements.get(3), expectedStatement);
+    }
+
+    @Test
+    public void shouldExceptionOnMissingSemicolon() throws Exception {
+        Path cqlPath = getResourcePath("cql_bootstrap_missing_semicolon/bootstrap.cql");
+        try {
+            List<String> cqlStatements = CqlFileParser.getCqlStatementsFrom(cqlPath);
```
Comment (adamdougal) [inline src/test/java/uk/sky/cqlmigrate/CqlFileParserTest.java]: The assignment isn't necessary here

**Comment by adamdougal** [comment]
Sorry, I also forgot to ask if you could add something in the readme file about the comment syntax? Thanks!

**Pull request merged**